File: 46-permutations/46-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        int n=nums.length;
        List<List<Integer>>res=new ArrayList<>();
        int []vis=new int [n];
        List<Integer>temp=new ArrayList<>();
        solve(0,nums,res,temp,vis);
        return res;
    }
    public void solve(int ind,int[] nums,List<List<Integer>>res,List<Integer>temp,int []vis){
        if(temp.size()==nums.length){
            res.add(new ArrayList<>(temp));
        }
        for(int i=0;i<nums.length;i++){
            if(vis[i]==0){
                temp.add(nums[i]);
                vis[i]=1;
                solve(ind+1,nums,res,temp,vis);
                vis[i]=0;
                temp.remove(temp.size()-1);
            }
        }
    }
}

File: 46-permutations/46-permutations.cpp
class Solution {
public:
    vector<vector<int>>res;
    void solve(vector<int> &nums,vector<int> &vis,vector<int>&temp){
        if(temp.size()==nums.size()){
            res.push_back(temp);
            return;
        }
        for(int i=0;i<nums.size();i++){
            if(vis[i]==0){
                vis[i]=1;
                temp.push_back(nums[i]);
                solve(nums,vis,temp);
                vis[i]=0;
                temp.pop_back();
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        int n=nums.size();
        vector<int>vis(n,0);
        vector<int>temp;
        solve(nums,vis,temp);
        return res;
    }
};